pubtools/_pulp/tasks/delete.py

```diff
@@ -0,0 +1,709 @@
+
+        self.parser.add_argument(
+            "--repo",
+            help="remove content from these comma-seperated repositories ",
+            type=str,
+            action=SplitAndExtend,
+            split_on=",",
+        )
+
+        self.parser.add_argument(
+            "--advisory",
+            help="remove packages in the advisory",
+            type=str,
+        )
+
+        self.parser.add_argument(
+            "--file",
+            help="remove these comma-separated rpm, srpm, modulemd or iso file(s)",
+            type=str,
+            action=SplitAndExtend,
+            split_on=",",
+        )
+
+        self.parser.add_argument(
+            "--signing-key",
+            help="remove the content with these signing-keys(s)",
+            type=str,
+            action=SplitAndExtend,
+            split_on=",",
+        )
+
+    def fail(self, *args, **kwargs):
+        LOG.error(*args, **kwargs)
+        sys.exit(30)
+
+    def run(self):
+        repos = {}
+        rpm_f = f_return()
+        file_f = f_return()
+        module_f = f_return()
+        adv_f = f_return()
+        signing_keys = self.args.signing_key
+        repo_names = self.args.repo
+        advisory = self.args.advisory
+
+        if not (self.args.file or self.args.advisory):
+            self.fail("One of --file or --advisory is required")
+
+        # separates files into rpms, isos and modules
+        rpms, files, modules, _ = self.separate_files(self.args.file or [])
+
+        # collect skipped items
+        # TODO
+
+        # delete rpms
+        if rpms:
+            rpm_f = self.delete_rpms(rpms, signing_keys, repo_names)
+
+        # delete files
+        if files:
+            file_f = self.delete_files(files, repo_names)
+
+        # delete modules
+        if modules:
+            module_f = self.delete_modules(modules, signing_keys, repo_names)
+
+        # delete packages from advisory
+        if advisory:
+            adv_f = self.delete_from_advisory(advisory, repo_names)
+
+        # wait for the futures to resolve
+        # each future returns a list of futures for content removal on each repo
+        fs = f_zip(rpm_f, file_f, module_f, adv_f)
+        fs = [f for sublist in fs.result() if sublist for f in sublist]
+        for f in fs:
+            repo = f.result().repo
+            repos.setdefault(repo.id, repo)
+
+        # publish affected repos
+        if repos:
+            publish_fs = self.publish_with_cache_flush(sorted(repos.values()))
+
+            f_sequence(publish_fs).result()
+
+        # wait for any futures that were not resolved during the process
+        if self.to_await:
+            for f in self.to_await:
+                f.result()
+
+    def delete_from_advisory(self, advisory, repos):
+        pushitems = []
+        advisory_id = None
+
+        # get errata from source
+        LOG.info("Get items from source")
+        with Source.get(advisory) as source:
```

Comment:
Not sure about this.  This is going to query ET and get push items on the advisory, as it exists today.

But I'd guess one of the main use-cases of this command is to deal with the side-effects of an advisory *changing*.  Let's say an advisory originally has packages A, B, C, D and a push is done.  Then the advisory is modified so it only contains packages A, B.  That's not going to "unpush" packages C, D, so I think that scenario is a case where someone may want to use this command in order to remove whatever was pushed by an earlier version of the advisory.  But that's not going to work if the code is based around looking at the *current* version of the advisory to decide what should be done.

Does it work this way on the old code? Deleting only what's in the advisory as fetched from ET during the task?

I think the command should work without querying ET and should just remove whatever's in Pulp associated with the advisory.

Reply:
Makes sense. Pulp should be the source-of-truth. However, current implementation gets the items from advisory source, verify that input repos are in item.dest(remove others in item.dest), map items to the repos in the input and remove those elements from the mapped repos.
IIUC, it should get the advisory from Pulp, gather the packages from the advisory and then remove them from the requested repos? Currently, only packages from the advisory is removed. Should it be extended to remove all the items i.e. both packages and modules in the advisory pkglist?

Reply:
If the current implementation is initializing the list of files by querying ET, then I'm unsure if this needs to change.  Though I believe that behavior doesn't make sense, it's not meant to be in scope here to fix all existing problems with the command.  I'd like to review some past uses of delete-packages-from-adv and see why it was used, and judge whether the command really did what the users wanted it to.  So let me have a bit more time to reply to this.

About whether it should handle modules too.  If it were a new command, I'd say yes and it doesn't make sense that it omits modules.  But again it wasn't the intended scope of this piece of work to fix all existing problems with the command.  What's your opinion though?  It also depends on how difficult it would be.

Reply:
I think it shouldn't be much of a work. We can get the list of module nsvca from the erratum pkglist and pass it to the delete_modules method to take care of the removal.

Reply:
About whether it should figure out the files for deletion by querying ET or by only querying Pulp.

I tried looking into historical tasks, but there's not a lot of data either way since the command only seems to be used a couple of times per year on average.  Still, I did find one example which shows how the concept of querying ET is flawed:

- here's a delete task: https://url.corp.redhat.com/e050479
- an example what it queried from ET: delete aap-azure-ui-0.0.1-0.1.20220221.git0a32a2d.el8ap.noarch.rpm from ansible-automation-platform-2_DOT_1-for-rhel-8-x86_64-rpms
- but ET content today: https://url.corp.redhat.com/14a02fb
- shows different repo(s) for the RPMs

In that case it worked out OK, but it seems error-prone: if someone had updated the repos in ET before the delete command was run, then the delete command wouldn't have found the correct repos for deletion.  I doubt that anyone is aware of that, and even if it hasn't bitten the users yet, it surely will eventually.

Combined with my opinion that, if the code were changed to query Pulp instead of ET then it would likely be not significantly more complex or maybe even simpler... yeah, I think it'd be best to revise it so it's just driven by Pulp state and doesn't call out to ET at all.

Reply:
Thanks. I updated the delete advisory workflow to fetch the details from Pulp, gather the packages and modules name and then pass them to delete_rpms and delete_modules for deletion. Also, it supports deleting multiple advisories now instead of one as in the current implementation.